Javadocs Feature

Reaper-1709 opened the pull request:
1. Added see javadocs button in StatementBlock, ExpressionBlock, and PluginComponentBlock
2. Added Javadocs tab in Plugin Settings
3. Javadocs functionality for all components, and all methods is added and no other blocks supported as of now

Reaper-1709 committed:
Javadocs initial commit
1. Added see javadocs button in StatementBlock, ExpressionBlock, and PluginComponentBlock
2. Added Javadocs tab in Plugin Settings
3. Javadocs functionality for all components, and all methods is added and no other blocks supported as of now

Sal4iDev commented:
there is a lot of code that repeats itself. wouldn't it be better to replace various pieces of code with one field and other pieces of code with a method?

Reaper-1709 commented:
> there is a lot of code that repeats itself. wouldn't it be better to replace various pieces of code with one field and other pieces of code with a method?

I'd work on optimisation once the project is finished maybe, I still have some ToDos: add javadocs reference for other statements, expressions and add a config file for explicitly redirecting any method and class.

If you could point out the references where code can be summed up into a method, I'd really appreciate that.

Reaper-1709 commented:
Please do not merge the PR yet, I mistakenly forgot to commit JavadocsManager.java

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/project/Project.java:
```diff
@@ -144,15 +150,248 @@ protected void updateItem(MavenModule item, boolean empty) {
+        javadocsUrlListView.getItems().clear();
+        javadocsUrlListView.getItems().add(DEFAULT_JAVADOCS_URL);
+        javadocsUrlListView.getSelectionModel().select(DEFAULT_JAVADOCS_URL);
+        javadocsUrlListView.getItems().add("https://docs.oracle.com/en/java/javase/21/docs/api/java.base/");
```
vb plugins use java 17

Reaper-1709 replied:
Thanks for the info, I'll change the javadocs url. Any other changes you'd like me to make?

OfficialDonut replied:
lmk when you commit JavadocsManager so i can test it

Reaper-1709 committed:
Changed java javadocs version to 1.17

Reaper-1709 committed:
Javadocs third commit:
1. roll-backed VisualBukkitApp.java due to presence of unnecessary changes
2. Commited JavadocsManager.java

Reaper-1709 committed:
Merge remote-tracking branch 'origin/master'

Reaper-1709 commented:
@OfficialDonut looks good to me now, you can test it.

OfficialDonut commented:
the JavadocsWindow class is also missing

Reaper-1709 commented:
> the JavadocsWindow class is also missing

That is an unnecessary one as well, please remove any code that mentions that class.

Reaper-1709 committed:
reverted VisualBukkitApp (hopefully)

Reaper-1709 committed:
Javadocs finalization:
- Added javadocs url for all statements and expressions
- Added remap system to explicitly redirect to a different javadocs page for a given method
- some optimizations in Project.java

Reaper-1709 commented:
I think it is fully completed now.

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/blocks/ExpressionBlock.java:
```diff
@@ -34,7 +39,119 @@ public ExpressionBlock() {
-                new ActionMenuItem(VisualBukkitApp.localizedText("context_menu.delete"), e -> UndoManager.current().execute(this::delete)));
+                new ActionMenuItem(VisualBukkitApp.localizedText("context_menu.delete"), e -> UndoManager.current().execute(this::delete)),
+                new ActionMenuItem(VisualBukkitApp.localizedText("context_menu.javadocs"), e -> {
+                    switch (this) {
```
instead of this switch add an overridable open javadoc method to Block

Reaper-1709 replied:
Alright thanks for the idea I'll do it

Reaper-1709 replied:
But won't it be the same thing? Like the switch won't be removed from here even then right?

OfficialDonut replied:
`new ActionMenuItem(VisualBukkitApp.localizedText("context_menu.javadocs"), e -> openJavadoc())`

Reaper-1709 replied:
Got it

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/blocks/ExpressionBlock.java:
```diff
@@ -34,7 +39,119 @@ public ExpressionBlock() {
+                        case ExprConditionalExpression exprConditionalExpression ->
+                                VisualBukkitApp.openURI(URI.create("https://docs.oracle.com/javase/tutorial/java/nutsandbolts/if.html"));
+                        case ExprConsumer exprConsumer ->
+                                VisualBukkitApp.openURI(URI.create("https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html"));
```
java 17 not 8

Reaper-1709 replied:
I couldn't find many helpful or even like javadocs urls for stuff so I had put in whatever best I could. I'll change this one don't worry.

Reaper-1709 committed:
Javadocs optimization:
- Added an overridable method as suggested
- removed switches from StatementBlock.java and ExpressionBlock.java
- changed java 8 urls to java 17 ones

Reaper-1709 commented:
It is now optimized as instructed.

Reaper-1709 committed:
Javadocs optimization:
- commit revised JavadocsManager.java

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/project/JavadocsManager.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    public static void getCompJavaDocs(PluginComponent pluginComponent) {
+        switch (pluginComponent.getBlockType().orElse("Unknown")) {
```
the plugin component classes should override openJavadocs() instead of this switch

OfficialDonut commented:
this doesnt compile

OfficialDonut commented on VB-Application/src/main/resources/lang/gui.properties:
```diff
@@ -51,10 +62,13 @@ label.disabled=Disabled
+label.javadocs=Javadocs
 label.maven=Maven
 label.modules=Modules
+label.no_javadocs = No packages found
```
```suggestion
label.no_javadocs=No packages found
```

OfficialDonut commented on VB-Application/src/main/resources/lang/gui_en_US.properties:
```diff
@@ -29,6 +29,7 @@ context_menu.paste_before=Paste Before
 context_menu.paste_inside=Paste Inside
 context_menu.pin=Pin
 context_menu.unpin=Unpin
+context_menu.javadocs = See Javadocs
```
```suggestion
context_menu.javadocs=See Javadocs
```

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/blocks/definitions/bungee/StatBungeeCordConnect.java:
```diff
@@ -1,19 +1,27 @@
+
 @BlockDefinition(id = "stat-bungeecord-connect", name = "BungeeCord Connect", description = "Connects a player to a BungeeCord server")
-public class StatBungeeCordConnect extends StatementBlock {
+public class  StatBungeeCordConnect extends StatementBlock {
```
```suggestion
public class StatBungeeCordConnect extends StatementBlock {
```

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/project/JavadocsManager.java:
```diff
@@ -0,0 +1,97 @@
+                getJavadocs("org.bukkit.event.inventory.InventoryClickEvent", "");
+                break;
+            case "comp-function":
+                getJavadocs("java.util.function.Function", "");
```
vb functions arent related to java.util.function.Function

Reaper-1709 replied:
So should I leave it empty or are vb functions related to something else in javadocs?

OfficialDonut replied:
empty

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/project/JavadocsManager.java:
```diff
@@ -0,0 +1,97 @@
+                getJavadocs("java.util.function.Function", "");
+                break;
+            case "comp-procedure":
+                getJavadocs("java.util.function.Supplier", "");
```
vb procedures arent related to java.util.function.Supplier

Reaper-1709 replied:
So should I leave it empty or are vb functions related to something else in javadocs?

OfficialDonut replied:
empty

OfficialDonut commented on VB-Application/src/main/java/com/gmail/visualbukkit/blocks/definitions/core/ExprColoredString.java:
```diff
@@ -1,18 +1,26 @@
 
+    @Override
+    public void openJavadocs() {
+        VisualBukkitApp.openURI(URI.create("https://jd.papermc.io/paper/1.21.4/org/bukkit/ChatColor.html"));
```
add a constant for the base url "https://jd.papermc.io/paper/1.21.4" or a utility method or something to avoid hardcoding it in a million places

Reaper-1709 committed:
Suggested changes implemented

Update StatBungeeCordConnect.java
Update gui.properties

Co-Authored-By: OfficialDonut <donut42915@gmail.com>

Reaper-1709 committed:
Add Javadocs integration for various components

Reaper-1709 committed:
Update .gitignore

Reaper-1709 committed:
Merge remote-tracking branch 'origin/master'

OfficialDonut commented:
If you click "see javadocs" on a method block where the class/method hasnt been selected it throws an npe
```
java.lang.NullPointerException: Cannot invoke "com.gmail.visualbukkit.reflection.ClassInfo.getName()" because the return value of "com.gmail.visualbukkit.blocks.parameters.ClassParameter.getValue()" is null
	at com.gmail.visualbukkit.project.JavadocsManager.getStatJavadocs(JavadocsManager.java:17)
	at com.gmail.visualbukkit.blocks.definitions.core.StatMethod.openJavadocs(StatMethod.java:29)
	at com.gmail.visualbukkit.blocks.StatementBlock.lambda$new$12(StatementBlock.java:37)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:232)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:189)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.controls/javafx.scene.control.MenuItem.fire(MenuItem.java:459)
	at javafx.controls/com.sun.javafx.scene.control.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1415)
	at javafx.controls/com.sun.javafx.scene.control.ContextMenuContent$MenuItemContainer.lambda$createChildren$12(ContextMenuContent.java:1368)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:247)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:232)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:189)
	at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.graphics/javafx.scene.Scene$MouseHandler.process(Scene.java:3984)
	at javafx.graphics/javafx.scene.Scene.processMouseEvent(Scene.java:1890)
	at javafx.graphics/javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2708)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:411)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:301)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$2(GlassViewEventHandler.java:450)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:424)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:449)
	at javafx.graphics/com.sun.glass.ui.View.handleMouseEvent(View.java:551)
	at javafx.graphics/com.sun.glass.ui.View.notifyMouse(View.java:937)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:185)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

Reaper-1709 committed:
Add error handling for Javadocs opening in ExprField, ExprMethod, and StatMethod

Reaper-1709 commented:
Should be fixed now, along with similar errors which could have occoured with ExprMethod and ExprField. Thanks a lot for pointing out these things and sorry for my lack of testing skills.

Reaper-1709 committed:
Fixed CompCommand and CompEventListener javadocs url generation errors
